File: 210-course-schedule-ii/course-schedule-ii.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        adj = {i: [] for i in range(numCourses)}
        indegree = [0] * numCourses
        for src, dest in prerequisites:
            adj[src].append(dest)
            indegree[dest] += 1
        
        q = [crs for crs in range(numCourses) if indegree[crs] == 0]
        res = []
        while q:
            cur = q.pop(0)
            for nei in adj[cur]:
                indegree[nei] -= 1
                if indegree[nei] == 0:
                    q.append(nei)
            res.append(cur)
        
        return res[::-1] if len(res) == numCourses else []
                

File: 210-course-schedule-ii/course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        unordered_map<int, vector<int>> adj;
        vector<int> indegree(numCourses, 0);
        queue<int> q;
        vector<int> res;
        for(auto pre: prerequisites) {
            int src = pre[0];
            int dest = pre[1];
            adj[src].push_back(dest);
            indegree[dest]++;
        }
        for(int i=0; i<numCourses; i++) {
            if(indegree[i]==0) q.push(i);
        }
        int finished = 0;
        while(!q.empty()) {
            int cur = q.front();
            q.pop();
            for(int nei: adj[cur]) {
                indegree[nei]--;
                if(indegree[nei] == 0) {
                    q.push(nei);
                }
            }
            finished++;
            res.push_back(cur);
        }
        reverse(res.begin(), res.end());
        if(finished == numCourses) {
            return res;
        } else {
            return {};
        }
    }
};